### PR TITLE
Prefixing settings with POOTLE_ and handling needed deprecations

### PR DESCRIPTION
**Pull request by dwaynebailey**
This is an attempt to address #3757 and it includes the following:

1. An approach to deprecation using Django checks framework and a 96-deprecation.conf file.
2. Migrating some settings to POOTLE_ and adding them to the deprecation settings.

Deprecation will mostly give info if there are deprecated settings.  It will warn if you haven't set the new setting but will copy the old setting to the new.  If it is still unchanged when we hit the removal date then it raises a critical error.  Pootle hangs then, it should exit nicely, but it achieves the same aim... Pootle doesn't run.

I haven't documented the deprecation rationale yet. But my thinking is to deprecate as fast as possible.

### CODE REVIEW

**Comment by dwaynebailey** [comment]
@translate/core this is ready for your review.  I think I've picked up all the comments from earlier interactions with reviewers.

File: pootle/apps/pootle_misc/context_processors.py [hunk 1]
```diff
@@ -34,9 +34,9 @@ def pootle_context(request):
     #FIXME: maybe we should expose relevant settings only?
     return {
         'settings': {
-            'TITLE': settings.TITLE,
+            'POOTLE_TITLE': settings.POOTLE_TITLE,
             'POOTLE_INSTANCE_ID': settings.POOTLE_INSTANCE_ID,
-            'CAN_CONTACT': settings.CAN_CONTACT and settings.CONTACT_EMAIL,
+            'POOTLE_CONTACT_ENABLED': settings.POOTLE_CONTACT_ENABLED and settings.POOTLE_CONTACT_EMAIL,
```
Comment (julen) [inline pootle/apps/pootle_misc/context_processors.py]: Nit: mind the line length.

File: pootle/settings.py [hunk 2]
```diff
@@ -10,9 +10,78 @@
...
+
 WORKING_DIR = os.path.abspath(os.path.dirname(__file__))
 
 
+DEPRECATIONS = [
+    # old, new, deprecated_version, removed_version
+    ('TITLE', 'POOTLE_TITLE', '2.7', '2.8'),
+    ('CAN_CONTACT', 'POOTLE_CONTACT_ENABLED', '2.7', '2.8'),
+    ('CONTACT_EMAIL', 'POOTLE_CONTACT_EMAIL', '2.7', '2.8'),
+    ('POOTLE_REPORT_STRING_ERRORS_EMAIL', 'POOTLE_CONTACT_REPORT_EMAIL', '2.7', '2.8'),
```
Comment (julen) [inline pootle/settings.py]: Line length.

File: pootle/settings.py [hunk 3]
```diff
@@ -10,9 +10,78 @@
...
+            continue
+
+        # Old setting is present and new setting is not defined:
+        # - Warn and copy
+        # - Fail hard if its too old
+        if hasattr(settings, old) and not hasattr(settings, new):
+            from pootle.__version__ import ver
+            if ver >= tuple(remove_ver.split(".")):
+                errors.append(checks.Critical(
+                    ("Setting %s is deprecated and was removed in Pootle %s." % (old, remove_ver)),
```
Comment (julen) [inline pootle/settings.py]: Line length.

File: pootle/settings.py [hunk 4]
```diff
@@ -10,9 +10,78 @@
...
+                    hint=("Use %s instead." % new),
+                    id="pootle.W002",
+                ))
+            else:
+                errors.append(checks.Warning(
+                    ("Setting %s is deprecated and will be removed in Pootle %s." % (old, remove_ver)),
```
Comment (julen) [inline pootle/settings.py]: Line length.

File: pootle/settings.py [hunk 5]
```diff
@@ -10,9 +10,78 @@
...
+                id="pootle.I002",
+            ))
+            continue
+
+        # Old setting is present and new setting is not defined:
+        # - Warn and copy
+        # - Fail hard if its too old
+        if hasattr(settings, old) and not hasattr(settings, new):
+            from pootle.__version__ import ver
+            if ver >= tuple(remove_ver.split(".")):
```
Comment (unho) [inline pootle/settings.py]: @dwaynebailey Please note that this might not work as desired: `(2, 5, 1) > '2.6'.split('.')` returns `True`.

Comment (unho) [inline pootle/settings.py]: @dwaynebailey The following might work as desired and is also simpler:

```python
            from pootle.__version__ import sver
            if sver >= remove_ver:
```

Comment (unho) [inline pootle/settings.py]: @dwaynebailey Sorry, forget my initial comment. Your code is fine, I was just testing a different code. Anyway the second comment still applies.

Comment (unho) [inline pootle/settings.py]: @dwaynebailey Sorry again, just tested your code with `(2, 6, 0) >= tuple('2.6'.split('.'))` and is returning `False`, so it doesn't work as expected. Instead `"2.6.0" >= "2.6"` returns `True` as desired. So please change those two lines of code to use instead what is on the second comment.

**Comment by unho** [comment]
@dwaynebailey Just wondering if this new deprecation mechanism is something we need to add to the releasing instructions.

File: pootle/settings.py (hunk 5 above)
Comment (dwaynebailey) [inline pootle/settings.py]: Well spotted but not 100% right since "2.10.0" < "2.6.0" but not so with (2, 10, 0) and (2, 6, 0). I found a better way.

**Comment by dwaynebailey** [comment]
@unho deprecation should happen when the code lands, not at release time.

**Comment by dwaynebailey** [comment]
@translate/core OK I've made adjustments based on previous feedback.

1. I've moved the checking code out of ``settings.py`` and into ``pootle.core.utils.deprecation``
2. We're now making use of ``AppConfig.ready()`` with a FIXME for when we move to Django 1.8
3. The minor nits have all landed.

File: docs/developers/deprecation.rst [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+Deprecation
+===========
+
+From time to time features, commands, configurations will be deprecated.  We
+deprecate and manage backward compatibility within the following guidelines:
+
+1. Our priority is the movement of Pootle development forward.  Thus:
+
+   1. We don't want to have to maintain backward compatability for too long as
```
Comment (julen) [inline docs/developers/deprecation.rst]: *compatibility*

File: docs/developers/deprecation.rst [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      it hampers forward mobility.
+   2. We won't maintain backward compatibility if that prevents or impacts the
+      needs of the new feature, refactoring, etc.
+   3. We won't maintain backward compatibility if the cost of that far
+      outweights the effort of reconfiguring Pootle.
+
+2. We don't want there to be major distruptions that we can avoid with point
```
Comment (julen) [inline docs/developers/deprecation.rst]: *disruptions*

File: docs/developers/deprecation.rst [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+   release.  That is it shuouldn't be painful as we shift features.
```
Comment (julen) [inline docs/developers/deprecation.rst]: *shouldn't*

File: docs/developers/deprecation.rst [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+2. If it is **obsolete or replaced** with an equivalent then drop with no
+   fanfare and add settings to the deprecation infrastructure so that an admin
+   will remove settings from their settings files.  Add to release notes if
+   needed.
+3. If it has been **renamed**.  Put that in the release notes and allow
+   fallback for one version.  Use deprecation infrastructure for settings to
+   allow old settings to continue to work until the N+1 release.  After that
+   its a hard failure.  For commands simply rename.
+4. If things **changed**.  For settings put that in release notes and do a hard
+   failure to ensure that admins will reconfigured.  For commands, just put
```
Comment (julen) [inline docs/developers/deprecation.rst]: *reconfigure*

File: docs/developers/deprecation.rst [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+   those notes in the release notes and in the command features.
+5. If **removed**. Put in release notes.  For settings choose either hard or
+   soft failure depending on whether something needs to be done by the admin.
+   Put in release notes together with a guide on how to work around the missing
+   feature if its possible. for commands simply make sure they are highlightd
```
Comment (julen) [inline docs/developers/deprecation.rst]: *it's*, *highlighted*

File: docs/developers/deprecation.rst [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+   as removed in the release notes.
+
+
+Implementing a deprecated settings
```
Comment (julen) [inline docs/developers/deprecation.rst]: singular, *setting*

**Comment by julen** [comment]
I only miss the renamed/deprecated settings in the 2.7 release notes as described in the new deprecation policy this PR includes. Other than that GTM.

**Pull request merged**